zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -27,16 +33,38 @@ pub struct Chain {
     sapling_nullifiers: HashSet<sapling::Nullifier>,
     orchard_nullifiers: HashSet<orchard::Nullifier>,
     partial_cumulative_work: PartialCumulativeWork,
+    history_tree: HistoryTree,
 }
 
 impl Chain {
+    /// Create a new empty non-finalized chain with the given history tree.
+    ///
+    /// The history tree must contain the history of the previous (finalized) blocks.
+    pub fn new(history_tree: HistoryTree) -> Self {
+        Chain {
+            blocks: Default::default(),
+            height_by_hash: Default::default(),
+            tx_by_hash: Default::default(),
+            created_utxos: Default::default(),
+            spent_utxos: Default::default(),
+            sprout_anchors: Default::default(),
+            sapling_anchors: Default::default(),
+            sprout_nullifiers: Default::default(),
+            sapling_nullifiers: Default::default(),
+            orchard_nullifiers: Default::default(),
+            partial_cumulative_work: Default::default(),
+            history_tree,
+        }
+    }
+
     /// Push a contextually valid non-finalized block into a chain as the new tip.
     #[instrument(level = "debug", skip(self, block), fields(block = %block.block))]
-    pub fn push(&mut self, block: PreparedBlock) {
+    pub fn push(&mut self, block: PreparedBlock) -> Result<(), HistoryTreeError> {
```

Comment:
API usability: let's use a more general error type `ValidateContextError`
```suggestion
    pub fn push(&mut self, block: PreparedBlock) -> Result<(), ValidateContextError> {
```